File: Core kernel Yaiwes/Backend watchdog workflow adaptativo/sandbox-runtime/iii-sandbox/test/sdk/queue.test.ts
import { describe, it, expect, vi, beforeEach } from "vitest";
import { QueueManager } from "../../packages/sdk/src/queue.js";
import { HttpClient } from "../../packages/sdk/src/client.js";

describe("QueueManager", () => {
  let mockClient: HttpClient;
  let queue: QueueManager;

  beforeEach(() => {
    mockClient = {
      get: vi.fn(),
      post: vi.fn(),
      del: vi.fn(),
      stream: vi.fn(),
    } as unknown as HttpClient;
    queue = new QueueManager(mockClient, "sbx_test123");
  });

  describe("submit", () => {
    it("calls correct endpoint with command", async () => {
      const job = {
        id: "job_abc",
        sandboxId: "sbx_test123",
        command: "echo hello",
        status: "pending",
        retries: 0,
        maxRetries: 3,
        createdAt: Date.now(),
      };
      (mockClient.post as any).mockResolvedValue(job);

      const result = await queue.submit("echo hello");

      expect(mockClient.post).toHaveBeenCalledWith(
        "/sandbox/sandboxes/sbx_test123/exec/queue",
        { command: "echo hello" },
      );
      expect(result).toEqual(job);
    });

    it("passes maxRetries and timeout options", async () => {
      (mockClient.post as any).mockResolvedValue({ id: "job_abc" });

      await queue.submit("ls", { maxRetries: 5, timeout: 30 });

      expect(mockClient.post).toHaveBeenCalledWith(
        "/sandbox/sandboxes/sbx_test123/exec/queue",
        { command: "ls", maxRetries: 5, timeout: 30 },
      );
    });
  });

  describe("status", () => {
    it("calls correct endpoint", async () => {
      const job = { id: "job_abc", status: "completed" };
      (mockClient.get as any).mockResolvedValue(job);

      const result = await queue.status("job_abc");

      expect(mockClient.get).toHaveBeenCalledWith(
        "/sandbox/queue/job_abc/status",
      );
      expect(result).toEqual(job);
    });
  });

  describe("cancel", () => {
    it("calls correct endpoint", async () => {
      (mockClient.post as any).mockResolvedValue({ cancelled: "job_abc" });

      const result = await queue.cancel("job_abc");

      expect(mockClient.post).toHaveBeenCalledWith(
        "/sandbox/queue/job_abc/cancel",
      );
      expect(result).toEqual({ cancelled: "job_abc" });
    });
  });

  describe("dlq", () => {
    it("calls correct endpoint without limit", async () => {
      const dlqResult = { jobs: [], total: 0 };
      (mockClient.get as any).mockResolvedValue(dlqResult);

      const result = await queue.dlq();

      expect(mockClient.get).toHaveBeenCalledWith("/sandbox/queue/dlq");
      expect(result).toEqual(dlqResult);
    });

    it("calls correct endpoint with limit", async () => {
      (mockClient.get as any).mockResolvedValue({ jobs: [], total: 0 });

      await queue.dlq(10);

      expect(mockClient.get).toHaveBeenCalledWith(
        "/sandbox/queue/dlq?limit=10",
      );
    });
  });
});

File: _work/backend-watchdog-full-repair-02/src/iii-sandbox/packages/sdk-rust/tests/queue_test.rs
use std::sync::Arc;

use iii_sandbox::queue::{QueueManager, QueueSubmitOptions};
use iii_sandbox::{ClientConfig, HttpClient};

fn make_queue(url: &str) -> QueueManager {
    let client = Arc::new(HttpClient::new(ClientConfig {
        base_url: url.to_string(),
        token: None,
        timeout_ms: None,
    }).unwrap());
    QueueManager::new(client, "sbx-1".into())
}

#[tokio::test]
async fn test_submit_command_only() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/sandboxes/sbx-1/exec/queue")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "id": "job-1",
                "sandboxId": "sbx-1",
                "command": "echo test",
                "status": "queued",
                "result": null,
                "error": null,
                "retries": 0,
                "maxRetries": 3,
                "createdAt": 1000,
                "startedAt": null,
                "completedAt": null
            })
            .to_string(),
        )
        .create_async()
        .await;

    let q = make_queue(&server.url());
    let result = q.submit("echo test", None).await.unwrap();
    assert_eq!(result.id, "job-1");
    assert_eq!(result.command, "echo test");
    assert_eq!(result.status, "queued");
    mock.assert_async().await;
}

#[tokio::test]
async fn test_submit_with_options() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/sandboxes/sbx-1/exec/queue")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "id": "job-2",
                "sandboxId": "sbx-1",
                "command": "python train.py",
                "status": "queued",
                "result": null,
                "error": null,
                "retries": 0,
                "maxRetries": 5,
                "createdAt": 1000,
                "startedAt": null,
                "completedAt": null
            })
            .to_string(),
        )
        .create_async()
        .await;

    let q = make_queue(&server.url());
    let result = q
        .submit(
            "python train.py",
            Some(QueueSubmitOptions {
                max_retries: Some(5),
                timeout: Some(60000),
            }),
        )
        .await
        .unwrap();
    assert_eq!(result.id, "job-2");
    assert_eq!(result.max_retries, 5);
    mock.assert_async().await;
}

#[tokio::test]
async fn test_status() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("GET", "/sandbox/queue/job-1/status")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "id": "job-1",
                "sandboxId": "sbx-1",
                "command": "echo test",
                "status": "completed",
                "result": {"exitCode": 0, "stdout": "test\n", "stderr": "", "duration": 0.01},
                "error": null,
                "retries": 0,
                "maxRetries": 3,
                "createdAt": 1000,
                "startedAt": 1001,
                "completedAt": 1002
            })
            .to_string(),
        )
        .create_async()
        .await;

    let q = make_queue(&server.url());
    let result = q.status("job-1").await.unwrap();
    assert_eq!(result.status, "completed");
    assert!(result.result.is_some());
    assert_eq!(result.result.unwrap().exit_code, 0);
    mock.assert_async().await;
}

#[tokio::test]
async fn test_cancel() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/queue/job-1/cancel")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(serde_json::json!({"cancelled": "job-1"}).to_string())
        .create_async()
        .await;

    let q = make_queue(&server.url());
    let result = q.cancel("job-1").await.unwrap();
    assert_eq!(result.cancelled, "job-1");
    mock.assert_async().await;
}

#[tokio::test]
async fn test_dlq_no_limit() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("GET", "/sandbox/queue/dlq")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "jobs": [],
                "total": 0
            })
            .to_string(),
        )
        .create_async()
        .await;

    let q = make_queue(&server.url());
    let result = q.dlq(None).await.unwrap();
    assert!(result.jobs.is_empty());
    assert_eq!(result.total, 0);
    mock.assert_async().await;
}

#[tokio::test]
async fn test_dlq_with_limit() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("GET", "/sandbox/queue/dlq?limit=10")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "jobs": [{
                    "id": "job-fail",
                    "sandboxId": "sbx-1",
                    "command": "bad cmd",
                    "status": "failed",
                    "result": null,
                    "error": "command not found",
                    "retries": 3,
                    "maxRetries": 3,
                    "createdAt": 1000,
                    "startedAt": 1001,
                    "completedAt": 1002
                }],
                "total": 1
            })
            .to_string(),
        )
        .create_async()
        .await;

    let q = make_queue(&server.url());
    let result = q.dlq(Some(10)).await.unwrap();
    assert_eq!(result.jobs.len(), 1);
    assert_eq!(result.total, 1);
    assert_eq!(result.jobs[0].error, Some("command not found".to_string()));
    mock.assert_async().await;
}
